File: python/543.diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0
    
        stack = [(root, False)]
        ret = {None: -1}

        while stack:
            node, pr = stack.pop()

            if pr:
                left = ret[node.left]
                right = ret[node.right]
                res = max(res, 2 + left + right)
                ret[node] = max(left, right) + 1

            else:
                stack.append((node, True))

                if node.right:
                    stack.append((node.right, False))

                if node.left:
                    stack.append((node.left, False))

        return res

File: cpp/543.diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;

        stack<pair<TreeNode*, bool>> st;
        st.push({root, false});

        unordered_map<TreeNode*, int> ret;
        ret[NULL] = -1;

        while (!st.empty()) {
            auto [node, pr] = st.top(); st.pop();

            if (pr) {
                int left = ret[node->left];
                int right = ret[node->right];
                res = max(res, left + right + 2);
                ret[node] = max(left, right) + 1;

            } else {
                st.push({node, true});

                if (node->right) {
                    st.push({node->right, false});
                }

                if (node->left) {
                    st.push({node->left, false});
                }
            }
        }

        return res;
    }
};